src/tools/pkcs11_uri.c

```diff
@@ -0,0 +1,342 @@
+{
+	struct pkcs11_uri *uri = calloc(1, sizeof(struct pkcs11_uri));
+	return uri;
+}
+
+int
+parse_pkcs11_uri(const char *input_string, struct pkcs11_uri *result)
+{
+	size_t length = 0;
+	char *uri;
+	char *path, *query, *token;
+	int rv = 0;
+
+	if (!input_string || !result) {
+		return 1;
+	}
+
+	/* Check PKCS#11 URI scheme */
+	length = strlen(PKCS11_URI_SCHEME);
+	if (strlen(input_string) < length || strncmp(input_string, PKCS11_URI_SCHEME, length) != 0) {
+		return 1;
+	}
+
+	/* Divide path and query part*/
+	if ((uri = strdup(input_string)) == NULL) {
+		rv = 1;
+		fprintf(stderr, "Error when allocating memory\n");
+		goto end;
+	}
+	path = uri;
+	path = strtok(path, "?");
+	query = strtok(NULL, "?");
+	if (path == NULL) {
+		rv = 1;
+		goto end;
+	}
+	/* Skip PKCS#11 URI scheme*/
+	path = path + length;
+
+	/* parse path */
+	token = strtok(path, ";");
+	while (token != NULL) {
+		char *argument = NULL, **result_ptr = NULL;
+		int *result_len = NULL;
+		int max_len = -1;
+		int id = get_attr(path_attr, token);
+		if (id == -1) {
+			rv = 1;
+			fprintf(stderr, "Invalid PKCS#11 uri attribute\n");
+			goto end;
+		}
+		argument = token + strlen(path_attr[id].name) + 1;
+
+		switch (id) {
+		case PKCS11_ID:
+			result_ptr = &result->id;
+			result_len = &result->id_len;
+			break;
+		case PKCS11_LIB_DESCRIPTION:
+			rv = 1;
+			fprintf(stderr, "PKCS#11 library description not supported\n");
+			goto end;
+			break;
+		case PKCS11_LIB_MANUFACTURER:
+			rv = 1;
+			fprintf(stderr, "PKCS#11 manufacturer ID not supported\n");
+			goto end;
+			break;
+		case PKCS11_LIB_VERSION:
+			rv = 1;
+			fprintf(stderr, "PKCS#11 library version not supported\n");
+			goto end;
+			break;
+		case PKCS11_MANUFACTURER:
+			result_ptr = &result->token_manufacturer;
+			max_len = 32;
+			break;
+		case PKCS11_MODEL:
+			result_ptr = &result->token_model;
+			max_len = 16;
+			break;
+		case PKCS11_OBJECT:
+			result_ptr = &result->object;
+			break;
+		case PKCS11_SERIAL:
+			result_ptr = &result->serial;
+			max_len = 16;
+			break;
+		case PKCS11_SLOT_DESCRIPTION:
+			result_ptr = &result->slot_description;
+			break;
+		case PKCS11_SLOT_ID:
+			result_ptr = &result->slot_id;
+			break;
+		case PKCS11_SLOT_MANUFACTURER:
+			result_ptr = &result->slot_manufacturer;
+			break;
+		case PKCS11_TOKEN:
+			result_ptr = &result->token_label;
+			max_len = 32;
+			break;
+		case PKCS11_TYPE:
+			result_ptr = &result->type;
+			break;
+		}
+		if (parse_string(argument, result_ptr, result_len, max_len) != 0) {
+			rv = 1;
+			goto end;
+		}
+		token = strtok(NULL, ";");
+	}
+
+	/* parse query */
+	if (query == NULL) {
+		goto end;
+	}
+	token = strtok(query, ";");
+	while (token != NULL) {
+		char *argument = NULL, **result_ptr = NULL;
+		int id = get_attr(query_attr, token);
+		if (id == -1) {
+			rv = 1;
+			fprintf(stderr, "Invalid PKCS#11 URI query\n");
+			goto end;
+		}
+		argument = token + strlen(query_attr[id].name) + 1;
+
+		switch (id) {
+		case PKCS11_ID:
```

Comment:
```suggestion
		case PKCS11_PIN_SOURCE:
```
same in the cases below